packages/commonwealth/client/scripts/controllers/server/sessionSigners/solana.ts

```diff
@@ -0,0 +1,119 @@
+  signers: Record<string, solw3.Keypair> = {};
+  private auths: Record<number, { payload: SessionPayload, signature: string }> = {};
+
+  getAddress(chainId: string): string | null {
+    return this.signers[chainId].publicKey.toBase58();
+  }
+
+  hasAuthenticatedSession(chainId: string): boolean {
+    this.getOrCreateSigner(chainId);
+    return this.signers[chainId] !== undefined && this.auths[chainId] !== undefined;
+  }
+
+  async getOrCreateAddress(chainId: string): Promise<string> {
+    return (await this.getOrCreateSigner(chainId)).publicKey.toBase58();
+  }
+
+  async authSession(chainId: string, payload: SessionPayload, signature: string) {
+    // TODO: verify signature is valid, as below in sign()
+    // TODO: verify payload datetime is valid
+    if (payload.address !== this.getAddress(chainId)) {
+      throw new Error(`Invalid auth: ${payload.address} vs. ${this.getAddress(chainId)}`);
+    }
+
+    this.auths[chainId] = { payload, signature };
+
+    const authStorageKey = `CW_SESSIONS-solana-${chainId}-auth`;
+    localStorage.setItem(authStorageKey, JSON.stringify(this.auths[chainId]));
+  }
+
+  private async getOrCreateSigner(chainId: string): Promise<solw3.Keypair> {
+    if (this.signers[chainId] !== undefined) {
+      return this.signers[chainId];
+    }
+    const storageKey = `CW_SESSIONS-solana-${chainId}`;
+    try {
+      const storage = localStorage.getItem(storageKey);
+      const { privateKey }: { privateKey: string } = JSON.parse(storage);
+      this.signers[chainId] = solw3.Keypair.fromSecretKey(bs58.decode(privateKey));
+
+      const authStorageKey = `CW_SESSIONS-solana-${chainId}-auth`
+      const auth = localStorage.getItem(authStorageKey);
+      if (auth !== null) {
+        const { payload, signature }: { payload: SessionPayload, signature: string } = JSON.parse(auth);
+        if (payload.address === this.getAddress(chainId)) {
+          this.auths[chainId] = { payload, signature };
+          // TODO: verify signature is valid, as below in sign()
+          // TODO: verify payload is not expired
+        } else {
+          console.log("Invalid auth:", payload.address, this.getAddress(chainId));
+        }
+      }
+    } catch (err) {
+      console.log("Could not restore previous session", err);
+      this.signers[chainId] = solw3.Keypair.generate();
+      delete this.auths[chainId]
+      const privateKey = bs58.encode(this.signers[chainId].secretKey);
+      localStorage.setItem(storageKey, JSON.stringify({ privateKey }))
+    }
+    return this.signers[chainId];
+  }
+
+  async sign(chainId: string, call: string, args: Record<string, ActionArgument>):
+  Promise<{
+    session: Session,
+    action: Action,
+    hash: string
+  }> {
+    const signer = this.signers[chainId];
+    const sessionPayload = this.auths[chainId]?.payload;
+    const sessionSignature = this.auths[chainId]?.signature;
+    // TODO: verify payload is not expired
```

Comment:
what's the plan with verifying that the payload has not expired? do we want to display a message saying "the session has expired" and then get the user to sign a new session payload? seems like there is a bit of UX thinking to be done there

Reply:
We already have a dialog in PR2 to the effect of "please reauthenticate" so we can just reject the session from inside the sessionSigner and the reauthentication flow will take over automatically. I think we're okay not doing any verification for now (in this PR), but would like to add this in PR3/5.